File: Payment_FS/PaymentService.fs
namespace Payment_FS

module PaymentService_FS =
    type Payment = 
    | Cash of decimal
    | CreditCard of (string * decimal) /// (ccNum * amount)

    let createCashPayment amount =
        if amount >= 0m
        then Ok (Cash amount)
        else Error "Cash must be a positive amount"

    let processPayment (payment: Payment) =
        match payment with
        | Cash amt ->
            sprintf $"Cash payment of {amt} received"
        | CreditCard (ccNum, amt) ->
            sprintf $"Credit card payment of {amt} recieved via credit card {ccNum}"
    
    let processPaymentCash amount =
        let payment = createCashPayment amount
        match payment with
        | Ok p ->
            let result = processPayment p
            sprintf $"Cash payment processd: {result}"
        | Error msg ->
            sprintf $"Invalid cash payment: {msg}"


    let add num1 num2 = num1 + num2

File: Payment_CS/PaymentService.cs
using System.Text.RegularExpressions;

namespace Payment_CS;

public abstract class PaymentBase
{
	public decimal Amount { get; set; }
	public string Error { get; set; }
	public abstract bool IsValid();
	public abstract string GetPaymentSummary();
}
public class CashPayment : PaymentBase
{
	public CashPayment(decimal amount)
	{
		Amount = amount;
	}

	public override bool IsValid()
	{
		if (Amount >= 0)
			return true;

		Error = "Cash must be a positive amount";
		return false;
	}

	public override string GetPaymentSummary() =>
		$"Cash payment of {Amount} received";
}

public class CreditCardPayment : PaymentBase
{
	public const string ccRegex = @"\d\d\d\d-\d\d\d\d-\d\d\d\d-\d\d\d\d";
	public string CreditCardNumber { get; set; }
	public CreditCardPayment(string ccNum, decimal amount)
	{
		CreditCardNumber = ccNum;
		Amount = amount;
	}

	public override bool IsValid()
	{
		if (Regex.IsMatch(CreditCardNumber, ccRegex))
			return true;
		
		Error = $"Invalid credit card number: {CreditCardNumber}";
		return false;
	}

	public override string GetPaymentSummary() =>
		$"Credit card payment of {Amount} received via credit card {CreditCardNumber}";
}

public static class PaymentService_CS
{
	public static string ProcessPayment(PaymentBase payment)
	{
		if (payment is null)
		{
			return "Invalid payment";
		}

		if (payment.IsValid())
		{
			return $"Payment processed: {payment.GetPaymentSummary()}";
		}

		return $"Invalid payment: {payment.Error}";
	}

	public static string ProcessPaymentCash(decimal amount) =>
		ProcessPayment(new CashPayment(amount));

	public static string ProcessPaymentCreditCard(string ccNum, decimal amount) =>
		ProcessPayment(new CreditCardPayment(ccNum, amount));
}